### PR TITLE
BI-1101 - people-search-v3 support

### PR DESCRIPTION
**Pull request by yosemsweet**
Updating Kissr to work with the new People Search v3 API offered by kiss metrics. This requires a whole scale rewrite of the Kissr internals but the end result is a much more powerful system. We now provide the capability of building your own reports directly from Kissr rather than having to build them within the KM interface. Generating a report does require more apriori knowledge of the interface and your KM events and properties, but we've found the increased capability to be well worth it.

BI-1101 Fix report so it still works when reports is list and not matrix.

BI-1101 Create function that checks is 'identity' and 'columns' fields are present. Add file for testing, which will be deleted before merge.

Building out the interface for the new people_search_v3 report generation system. We can now generate json for the report target segment by using a combination of rules properties. Still need to test much of the edge cases and build out the columns. Also need to provide a mechanism for accessing KM events and properties to get their ids

Got options working for KissSegment and asJson. Also added some sample json files and bash files for testingand KM api reference

Think I have the basics of generating a report request json down. There are a lot of extras missing, and still need to do a lot of testing, but the heavy lifting is complete now

It actually submits the request and brings back data now! We have to do a bunch of work to get the columns labeled appropriately, and we don't expose most of the functionality available in the new KM api, and the documentation is poor, and the code needs a clean up pass. But we actually request a report from KM and it returns data (although we don't parse the data properly any more).

Automagically convert date based calculations from epoch timestamps to POSIXct datetimes

Added documentation for asJson - started docs for KissCalculation

BI-1101 Create functions to 1) pull full list of KM events 2) pull full list of KM properties and 3) pull the idex of a single property or event for use in kiss-report function.

Fixing documentation, cleaning up code, incremented version to 1.0.0.0

Fixed a typo in examples. Added documentation for asJson and names.KissReport

Fixed 2 bugs having to do with properties - KissCalculation.Property had the wrong type set, that's fixed now. And KissRule.Property needed to support a ComparisonString for specific ComparisonModes. Also fixed

Prettified json file

BI-1101 Insert date range into KissRule.Event and KissRule.Property, and correct asJson.KissRule to use date range when interval is fed to these functions.

Some minor code fixes for intervals with KissRules

Fixed a bug in kiss rule asJson (missing comma). Switched the time conversion to use dplyr to fix a warning about multiple assignment

BI-1101 Fix asJson function for Property attrbutes

update bash example so we aren't leaking API keys. Updated examples for KissRule.property

Handle empty reports

### CODE REVIEW

File: R/kiss-attributes.R [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+}
+
+# Outputs event/property index used as paramenter in KissReport function
+GetAttributeIndex <- function(attributeName,
+                              productId = "6581c29e-ab13-1030-97f2-22000a91b1a1"){
+  fullAttributeList <- GetAttributeList()
+
+  attributeIndex <- filter(fullAttributeList, display_name == attributeName &
+                             product_id == productId)
+
```
Comment (jonathan-marsan) [inline R/kiss-attributes.R]: Change filter to dplyr::filter

File: R/kiss-attributes.R [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  }
+}
+
+# Output event/property name from index number
+GetAttributeName <- function(attributeIndex,
+                              productId = "6581c29e-ab13-1030-97f2-22000a91b1a1",
+                             metricType = "event"){
+  fullAttributeList <- GetAttributeList()
+  metricType <- metricType
+  attributeName <- filter(fullAttributeList,index == attributeIndex &
```
Comment (jonathan-marsan) [inline R/kiss-attributes.R]: ditto above: Change filter to dplyr::filter

File: R/kiss-attributes.R [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+  if(nrow(attributeName) == 1){
+    return(attributeName$"name")
+  } else if(nrow(attributeName) > 1) {
+    stop(paste0("More than one attribute with that index.\n",
+                "Use GetAttributeList() for full list of attributes (index column)."))
+  } else {
+    stop(paste0("Attribute index not in table.\n",
+                "Check full attribute list by using GetAttributeList()."))
+  }
```
Comment (jonathan-marsan) [inline R/kiss-attributes.R:55]: Add following to error message:
Default looks for events only, first try looking for properties by specifying metricType = "property", otherwise...

File: R/kiss-rule.R [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+#' Generates json for a KissRule.
+#' @export
+asJson.KissRule.Event <- function(rule) {
+  template <- '
+  {
+    "type":"{{type}}",
+    "negate": {{negate}},
+    "event":{{event}},
+    "frequencyValue":"{{frequencyValue}}",
+    "frequencyOccurance":"{{frequencyOccurance}}",
```
Comment (jonathan-marsan) [inline R/kiss-rule.R:114]: Remove quotes from frequencyValue

**Comment by jonathan-marsan** [comment]
Fixed date range issue, but still need to clean up dates because the conversion does not seem to be working atm,
i.e. currently you feed the UTC time to the function and you get back the PDT time.

**Comment by jonathan-marsan** [comment]
Merging this since @yosemsweet on vacation.
Report itself works, if changes are needed we should create a different JIRA ticket.
Will close BI-1101 at same time.

**Comment by jonathan-marsan** [comment]
Not deleting branch to give Yosem chance to review.
